keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,707 @@
+
+However, there are a [number of use cases](#user-stories) that would benefit from Kubernetes nodes supporting swap. Hence, this proposal aims to add swap support to nodes in a controlled, predictable manner so that Kubernetes users can perform testing and provide data to continue building cluster capabilities on top of swap.
+
+## Motivation
+
+There are two distinct types of user for swap, who may overlap:
+- node administrators, who may want swap available for node-level performance tuning and stability/reducing noisy neighbour issues
+- application developers, who have written applications that would benefit from using swap memory
+
+There are hence a number of possible ways that one could envision swap use on a node.
+
+### Scenarios
+
+1. Swap is enabled only at the system level. The CRI does not permit user workloads to use swap. (This scenario is a prerequisite for the following use cases.)
+1. Swap is enabled at the node level. The CRI can be globally configured to permit user workloads scheduled on the node to use some quantity of swap.
+1. Swap is set on a per-workload basis. The CRI sets permitted swap utilization on each individual workload.
+
+This KEP will be limited in scope to the first two scenarios. The third can be addressed in a follow-up KEP. The enablement work that is in scope for this KEP will be necessary to implement the third scenario.
+
+
+### Goals
+
+- On Linux systems, when swap is provisioned and available, Kubelet can start up with swap on.
+- Configuration is available for CRI to set swap utilization available to Kubernetes workloads, defaulting to 0 swap.
+- Cluster administrators can enable and configure CRI swap utilization on a per-node basis.
+- Use of swap memory with both cgroupsv1 and cgroupsv2 is supported.
+
+### Non-Goals
+
+- Provisioning swap. Swap must already be available on the system.
+- Setting [swappiness]. This can already be set on a system-wide level outside of Kubernetes.
+- Allocating swap on a per-workload basis with accounting (e.g. pod-level specification of swap). If desired, this should be designed and implemented as part of a follow-up KEP. This KEP is a prerequisite for that work.
+- Supporting zram, zswap, or other memory types like SGX EPC. These could be addressed in a follow-up KEP, and are out of scope.
+
+[swappiness]: https://en.wikipedia.org/wiki/Memory_paging#Swappiness
+
+## Proposal
+
+We propose that, when swap is provisioned and available on a node, cluster administrators can configure the Kubelet and CRI such that:
+
+- The kubelet can start with swap on.
+- The CRI is updated such that by default, workloads will use 0 swap.
```

Comment:
If the applied setting is 0 and a Pod uses `tmpfs`-backed storage, can that tmpfs data go to swap?

Reply:
I believe the answer is no in the default case, but if a workload is using tmpfs and is permitted to use swap, that memory could be swapped out. We should include some tmpfs test scenarios.

Reply:
Added to GA criteria.